README.md
<p align="center" style="color: #343a40">
  <img src="https://cdn.rawgit.com/tkh44/emotion/master/emotion.png" alt="emotion" height="200" width="200">
  <h1 align="center">emotion</h1>
</p>
<p align="center" style="font-size: 1.2rem;">The Next Generation of CSS-in-JS</p>

[![npm version](https://badge.fury.io/js/emotion.svg)](https://badge.fury.io/js/emotion)
[![Build Status](https://travis-ci.org/tkh44/emotion.svg?branch=master)](https://travis-ci.org/tkh44/emotion)
[![codecov](https://codecov.io/gh/tkh44/emotion/branch/master/graph/badge.svg)](https://codecov.io/gh/tkh44/emotion)
![core gzip size](http://img.badgesize.io/https://unpkg.com/emotion/dist/DO-NOT-USE.min.js?compression=gzip&label=core%20gzip%20size)
![core size](http://img.badgesize.io/https://unpkg.com/emotion/dist/DO-NOT-USE.min.js?label=core%20size)
![react gzip size](http://img.badgesize.io/https://unpkg.com/react-emotion/dist/DO-NOT-USE.min.js?compression=gzip&label=react%20gzip%20size)
![react size](http://img.badgesize.io/https://unpkg.com/react-emotion/dist/DO-NOT-USE.min.js?label=react%20size)
[![slack](https://emotion.now.sh/badge.svg)](http://slack.emotion.sh/)

emotion is a high performance, lightweight css-in-js library. 
The core idea comes from Sunil Pai’s [glam](https://github.com/threepointone/glam) library and its philosophy is laid out [here](https://gist.github.com/threepointone/0ef30b196682a69327c407124f33d69a). 
The basic idea is simple.
You shouldn’t have to sacrifice runtime performance for good developer experience when writing CSS. emotion 
minimizes the runtime cost of css-in-js dramatically by parsing your styles with PostCSS during compilation instead of at runtime. 

-- [Introduction Article](https://medium.com/@tkh44/emotion-ad1c45c6d28b)

- [Install](https://github.com/tkh44/emotion/tree/master/docs/install.md)
- Demos
  - [emotion website](https://github.com/tkh44/emotion/tree/master/example) [[Demo Here](https://emotion.sh)]
  - [next-hnpwa-guide-kit](https://github.com/tkh44/next-hnpwa-guide-kit) [[Demo Here](https://hnpwa.life)]
  - **open a PR and add yours!**



- [Documentation](https://github.com/tkh44/emotion/tree/master/docs/index.md)

  - [babel-macros](https://github.com/tkh44/emotion/tree/master/docs/babel-macros.md)
  - [composes](https://github.com/tkh44/emotion/tree/master/docs/composes.md)
  - [css](https://github.com/tkh44/emotion/tree/master/docs/css.md)
  - [css-modules](https://github.com/tkh44/emotion/tree/master/docs/css-modules.md)
  - [css-prop](https://github.com/tkh44/emotion/tree/master/docs/css-prop.md)
  - [extract-static](https://github.com/tkh44/emotion/tree/master/docs/extract-static.md)
  - [font-face](https://github.com/tkh44/emotion/tree/master/docs/font-face.md)
  - [inject-global](https://github.com/tkh44/emotion/tree/master/docs/inject-global.md)
  - [install](https://github.com/tkh44/emotion/tree/master/docs/install.md)
  - [keyframes](https://github.com/tkh44/emotion/tree/master/docs/keyframes.md)
  - [media](https://github.com/tkh44/emotion/tree/master/docs/media.md)
  - [nested](https://github.com/tkh44/emotion/tree/master/docs/nested.md)
  - [objects](https://github.com/tkh44/emotion/tree/master/docs/objects.md)
  - [preact](https://github.com/tkh44/emotion/tree/master/docs/preact.md)
  - [props](https://github.com/tkh44/emotion/tree/master/docs/props.md)
  - [pseudo](https://github.com/tkh44/emotion/tree/master/docs/pseudo.md)
  - [ssr](https://github.com/tkh44/emotion/tree/master/docs/ssr.md)
  - [styled](https://github.com/tkh44/emotion/tree/master/docs/styled.md)
  - [styled-with-object](https://github.com/tkh44/emotion/tree/master/docs/styled-with-object.md)
  - [styling-any-component](https://github.com/tkh44/emotion/tree/master/docs/styling-any-component.md)
  - [theming](https://github.com/tkh44/emotion/tree/master/docs/theming.md)

- Ecosystem
  - [emotion-vue](https://github.com/egoist/emotion-vue)


CHANGELOG.md
## v9.0.0 (UNRELEASED)

**Emotion 9 is currently in beta and there may be breaking changes between versions**

Emotion 9 introduces [instances](https://emotion.sh/docs/create-emotion), jest-emotion, two new SSR methods and a brand new website!! Instances allow you to customize prefixing options, run plugins during CSS processing, set nonces for CSP and they're totally optional so if you don't need instances keep on using `emotion`. jest-emotion offers a better out of the box experience and has the potential for more features in the future. emotion-server has new methods for inlining CSS in HTML right before it's needed and adds support for streaming.

#### :rocket: Enhancement

* `jest-emotion`
  * [#495](https://github.com/emotion-js/emotion/pull/495) Add jest-emotion ([@mitchellhamilton](https://github.com/mitchellhamilton))
* `emotion-server`
  * [#448](https://github.com/emotion-js/emotion/pull/448) Add SSR streaming API and stuff ([@mitchellhamilton](https://github.com/mitchellhamilton))

#### Migrating from Emotion 8

1. Emotion no longer has `fontFace`, you can use `injectGlobal` to insert font faces instead.

```diff
-import { fontFace } from 'emotion'
+import { injectGlobal } from 'emotion'

-fontFace`
-  font-family: 'Oxygen';
-  font-style: normal;
-  font-weight: 400;
-  src: local('Oxygen Regular'), local('Oxygen-Regular'),
-    url(https://fonts.gstatic.com/s/oxygen/v6/qBSyz106i5ud7wkBU-FrPevvDin1pK8aKteLpeZ5c0A.woff2)
-      format('woff2');
-  unicode-range: U+0000-00ff, U+0131, U+0152-0153, U+02c6, U+02da, U+02dc,
-    U+2000-206f, U+2074, U+20ac, U+2212, U+2215;
-`
+injectGlobal`
+  @font-face {
+    font-family: 'Oxygen';
+    font-style: normal;
+    font-weight: 400;
+    src: local('Oxygen Regular'), local('Oxygen-Regular'),
+      url(https://fonts.gstatic.com/s/oxygen/v6/qBSyz106i5ud7wkBU-FrPevvDin1pK8aKteLpeZ5c0A.woff2)
+        format('woff2');
+    unicode-range: U+0000-00ff, U+0131, U+0152-0153, U+02c6, U+02da, U+02dc,
+      U+2000-206f, U+2074, U+20ac, U+2212, U+2215;
+  }
+`
```

2. If you used [jest-glamor-react](https://github.com/kentcdodds/jest-glamor-react) in emotion 8, you can switch to jest-emotion by installing it and changing your test setup as shown below.

```bash
npm install --save jest-emotion
```

```diff
-import { sheet } from 'emotion'
-import serializer from 'jest-glamor-react'
+import * as emotion from 'emotion'
+import { createSerializer } from 'jest-emotion'

-expect.addSnapshotSerializer(serializer(sheet))
+expect.addSnapshotSerializer(createSerializer(emotion))
```

3. Emotion doesn't automatically insert semicolons in styles now. We recommend using [Prettier](https://prettier.io/) to ensure that your styles are formatted correctly.
4. `ThemeProvider` and `withTheme` are now exported by `react-emotion` and `preact-emotion` instead of `emotion-theming`, you can install [`jscodeshift`](https://github.com/facebook/jscodeshift) globally, put the code below into a file named `transform.js` and run `jscodeshift PATH_YOU_WANT_TO_TRANSFORM` to update your imports.

```jsx
export default function transformer(file, api) {
  const j = api.jscodeshift
  let quote = 'single'
  return j(file.source)
    .find(j.ImportDeclaration, {
      source: { value: 'emotion-theming' }
    })
    .forEach(path => {
      // replace react-emotion with preact-emotion if you're using Preact
      path.node.source.value = 'react-emotion'
      if (path.node.source.raw.charAt(0) === '"') {
        quote = 'double'
      }
    })
    .toSource({ quote })
}
```

5. That's It!! You might want to check the more detailed list of breaking changes below just in case though.

#### :boom: Breaking Change

* `emotion`
  * `fontFace` is gone, it can be replaced with `injectGlobal` with a regular css `@font-face` rule.
  * The `registered` and `inserted` caches are no longer exports of emotion (they're now on the `caches` object which is exported by emotion) but these aren't documented and should never be relied on externally so this shouldn't cause a problem.
  * Semicolons are not automatically added in style blocks.
  * `StyleSheet` is no longer used for SSR, this is completely internal(except that emotion exports an instance of this as `sheet`) so it shouldn't cause any problems.
* `emotion-server`
  * `extractCritical` no longer returns a rules property, this was never documented so it shouldn't cause any problems.

## v9.0.0-2 (2018-01-17)

#### :rocket: Enhancement

* `create-emotion-styled`, `create-emotion`, `emotion-utils`, `emotion`, `react-emotion`
  * [#516](https://github.com/emotion-js/emotion/pull/516) Throw a nice error when using the styled shorthand without babel-plugin-emotion and remove duplication in component selector code. ([@mitchellhamilton](https://github.com/mitchellhamilton))

#### :bug: Bug Fix

* `create-emotion`, `emotion`
  * [#532](https://github.com/emotion-js/emotion/pull/532) Fix extra whitespace in cx. ([@mitchellhamilton](https://github.com/mitchellhamilton))

#### :memo: Documentation

* Other
  * [#533](https://github.com/emotion-js/emotion/pull/533) Reusable Media Queries with String Styles wasn't working. ([@ARChilton](https://github.com/ARChilton))
  * [#529](https://github.com/emotion-js/emotion/pull/529) Update examples in Readme. ([@divyanshu013](https://github.com/divyanshu013))
  * [#509](https://github.com/emotion-js/emotion/pull/509) fix Readme markdown list syntax. ([@ryota-murakami](https://github.com/ryota-murakami))
  * [#505](https://github.com/emotion-js/emotion/pull/505) Fixed jest-glamor-react -> jest-emotion upgrade description. ([@Andarist](https://github.com/Andarist))
* `site`
  * [#530](https://github.com/emotion-js/emotion/pull/530) Add search to docs. ([@mitchellhamilton](https://github.com/mitchellhamilton))
* `babel-plugin-emotion`
  * [#514](https://github.com/emotion-js/emotion/pull/514) Better example for hoist option. ([@Andarist](https://github.com/Andarist))

#### :house: Internal


example/src/main.js
import React from 'react'
import { render } from 'react-dom'
import { ThemeProvider, withTheme } from 'theming'
import styled from 'react-emotion'
import { css, fontFace, keyframes, injectGlobal } from 'emotion'
import Markdown from './markdown'
import Playground from './playground'
import logoUrl from '../../emotion.png'

const introExample = require('./blocks/intro.example')
const propsExample = require('./blocks/props.example')
const nestedExample = require('./blocks/nested.example')
const mediaExample = require('./blocks/media.example')
const anyComponentExample = require('./blocks/styling-any-component.example')
const pseudoExample = require('./blocks/pseudo.example')
const keyframesExample = require('./blocks/keyframes.example')
// const fontFaceExample = require('./blocks/font-face.example')
// const docMarkdown = require('./docs/index.md')
// const readmeMarkdown = require('../../README.md')

export const scope = {
  logoUrl,
  React,
  css,
  keyframes,
  styled,
  ThemeProvider,
  withTheme,
  require(name) {
    switch (name) {
      case 'emotion':
        return require('emotion')
      case 'emotion/react':
      case 'react-emotion':
        return require('react-emotion')
      case 'theming':
        return require('theming')
    }
  }
}

injectGlobal`
  html, body {
    font-family: -apple-system,
      BlinkMacSystemFont,
      "Segoe UI",
      "Roboto",
      "Roboto Light",
      "Oxygen",
      "Ubuntu",
      "Cantarell",
      "Fira Sans",
      "Droid Sans",
      "Helvetica Neue",
      sans-serif,
      "Apple Color Emoji",
      "Segoe UI Emoji",
      "Segoe UI Symbol";
    color: #495057;
    width: 100%;
    height: 100%;
    padding: 0;
    margin: 0;
  }
`

fontFace`
  font-family: 'Oxygen';
  font-style: normal;
  font-weight: 400;
  src: local('Oxygen Regular'), local('Oxygen-Regular'), url(https://fonts.gstatic.com/s/oxygen/v6/qBSyz106i5ud7wkBU-FrPevvDin1pK8aKteLpeZ5c0A.woff2) format('woff2');
  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2212, U+2215;
`

const theme = {
  white: '#f8f9fa',
  purple: '#8c81d8',
  gold: '#ffd43b'
}

const margin = (t, r, b, l) => {
  return () => css`
    margin-top: ${t};
    margin-right: ${r};
    margin-bottom: ${b};
    margin-left: ${l};
  `
}

const PlaygroundWrapper = styled.div`
  font-family: 'Oxygen', sans-serif;
  flex: 1;
  color: #343a40;
  background: #f8f9fa;

  & .inner {
    ${margin(0, 'auto', 0, 'auto')};
    width: calc(100% - 32px);
    max-width: 960px;

    @media (min-width: 960px) {
      width: 100%;
    }
  }

  & .header {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 100%;

    & img {
      display: block;
      width: 128px;
      height: 128px;
    }
  }
`

class App extends React.Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <PlaygroundWrapper>
          <div className="inner">
            <div className="header">
              <h1>
                <img src={logoUrl} alt="emotion" />
                emotion
              </h1>
              <p>The Next Generation of CSS-in-JS</p>
            </div>

            <Playground
              noRender={false}
              codeText={introExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/install.md')} />

            <Markdown markdown={require('../../docs/nested.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={nestedExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/pseudo.md')} />
            <Playground
              maxHeight={220}
              noRender={false}
              codeText={pseudoExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/media.md')} />
            <Playground
              maxHeight={220}
              noRender={false}
              codeText={mediaExample}
              scope={scope}
            />

            <Markdown
              markdown={require('../../docs/styling-any-component.md')}
            />
            <Playground
              maxHeight={220}
              noRender={false}
              codeText={anyComponentExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/props.md')} />
            <Playground
              noRender={false}
              codeText={propsExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/keyframes.md')} />
            <Playground
              maxHeight={600}
              noRender={false}
              codeText={keyframesExample}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/composes.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={require('./blocks/composes.example')}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/objects.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={require('./blocks/objects.example')}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/styled-with-object.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={require('./blocks/styled-with-object.example')}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/css-prop.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={require('./blocks/css-prop.example')}
              scope={scope}
            />

            <Markdown markdown={require('../../docs/theming.md')} />
            <Playground
              maxHeight={180}
              noRender={false}
              codeText={require('./blocks/theming.example')}
              scope={scope}
            />
          </div>
        </PlaygroundWrapper>
      </ThemeProvider>
    )
  }
}

render(<App />, document.getElementById('app'))
